common/transport/amqp/src/receiver_link.ts

```diff
@@ -0,0 +1,212 @@
+                this._fsm.transition('detached', callback, err);
+              });
+            } else {
+              this._fsm.transition('detached', callback, err);
+            }
+          },
+          '*': () => this._fsm.deferUntilTransition('detached')
+        }
+      }
+    });
+
+    this.on('removeListener', (eventName) => {
+      // stop listening for AMQP events if our consumers stop listening for our events
+      if (eventName === 'message' && this.listeners('message').length === 0) {
+        this._fsm.handle('detach');
+      }
+    });
+
+    this.on('newListener', (eventName) => {
+      // lazy-init AMQP event listeners
+      if (eventName === 'message') {
+        this._fsm.handle('attach');
+      }
+    });
+  }
+
+  detach(callback: (err?: Error) => void): void {
+    this._fsm.handle('detach', callback);
+  }
+
+  forceDetach(err?: Error): void {
+    this._fsm.handle('forceDetach', err);
+  }
+
+  attach(callback: (err?: Error) => void): void {
+    this._fsm.handle('attach', callback);
+  }
+
+  accept(message: Message, callback?: (err?: Error, result?: results.MessageCompleted) => void): void {
+    /*Codes_SRS_NODE_AMQP_RECEIVER_LINK_16_021: [The `accept` method shall throw if the `message` argument is falsy.]*/
+    if (!message) { throw new ReferenceError('Invalid message object.'); }
+    this._fsm.handle('accept', message.transportObj, callback);
+  }
+
+  complete(message: Message, callback?: (err?: Error, result?: results.MessageCompleted) => void): void {
+    /*Codes_SRS_NODE_AMQP_RECEIVER_LINK_16_015: [The `complete` method shall call the `accept` method with the same arguments (it is here for backward compatibility purposes only).]*/
+    this.accept(message, callback);
+  }
+
+  reject(message: Message, callback?: (err?: Error, result?: results.MessageRejected) => void): void {
+    /*Codes_SRS_NODE_AMQP_RECEIVER_LINK_16_021: [The `reject` method shall throw if the `message` argument is falsy.]*/
+    if (!message) { throw new ReferenceError('Invalid message object.'); }
+    this._fsm.handle('reject', message.transportObj, callback);
+  }
+
+  abandon(message: Message, callback?: (err?: Error, result?: results.MessageAbandoned) => void): void {
+    /*Codes_SRS_NODE_AMQP_RECEIVER_LINK_16_021: [The `abandon` method shall throw if the `message` argument is falsy.]*/
+    if (!message) { throw new ReferenceError('Invalid message object.'); }
+    this._fsm.handle('abandon', message.transportObj, callback);
+  }
+
+  private _removeListeners(): void {
+    this._linkObject.removeListener('detached', this._detachHandler);
+    this._linkObject.removeListener('message', this._messageHandler);
+    this._linkObject.removeListener('errorReceived', this._errorHandler);
+  }
+
+  private _safeCallback(callback: (err?: Error, result?: any) => void, error?: Error | null, result?: any): void {
+    if (callback) {
+      process.nextTick(() => callback(error, result));
```

Comment:
>process.nextTick(() => callback(error, result)); [](start = 6, length = 48)

same comments on _safeCallback that I made in some other file.

Reply:
same answer

---
In reply to: [139544184](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139544184) [](ancestors = 139544184)